java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,489 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sage.io;
+
+import sage.Sage;
+
+import java.io.*;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+
+public class RemoteSageFile implements SageFileSource
+{
+  private final String hostname;
+  private final String remoteFilename;
+  private final String transcodeMode;
+  private final boolean readonly;
+
+  private boolean activeFile = false;
+  private long currTotalSize = 0;
+  private long maxRemoteSize = 0;
+
+  private long remoteOffset = 0;
+  private Socket socket = null;
+  private DataOutputStream outStream = null;
+  private DataInputStream inStream = null;
+
+  public RemoteSageFile(String hostname, File file, boolean readonly) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, null);
+  }
+
+  public RemoteSageFile(String hostname, String name, boolean readonly) throws IOException
+  {
+    this(hostname, name, readonly, null);
+  }
+
+  public RemoteSageFile(String hostname, File file, boolean readonly, String transcodeMode) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, transcodeMode);
+  }
+
+  /**
+   *
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param transcodeMode Set the transcode mode.
+   * @throws IOException
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, String transcodeMode) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = transcodeMode;
+    this.readonly = readonly;
+    connect();
+  }
+
+  private void connect() throws IOException
+  {
+    if (socket != null)
+      return;
+
+    Socket newSocket = new Socket();
+    newSocket.connect(new InetSocketAddress(hostname, 7818));
+    newSocket.setSoTimeout(30000);
+    newSocket.setTcpNoDelay(true);
+    socket = newSocket;
+
+    outStream = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    inStream = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+
+    if (transcodeMode != null && transcodeMode.length() > 0)
+    {
+      outStream.write(("XCODE_SETUP " + transcodeMode + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      String response = Sage.readLineBytes(inStream);
+      if (!"OK".equals(response))
+        throw new java.io.IOException("Error with remote transcode setup for " + transcodeMode + " of: " + response);
+    }
+
+    outStream.write("OPENW ".getBytes(Sage.BYTE_CHARSET));
```

Comment:
If you change that behavior; just make sure you're opening it in the proper read/write mode since that's how read/write security is enforced in MediaServer.